### ISSUE TITLE
CreateInitialSysCatalogSnapshotTest.CreateInitialSysCatalogSnapshot fails in clang11 debug build with optimizations disabled

### ISSUE THREAD

**Issue by ttyusupov** [yugabyte/yugabyte-db#7774]
Jira Link: [DB-1604](https://yugabyte.atlassian.net/browse/DB-1604)
Steps to reproduce:
- `git checkout 21318956a2b5274bf3a1a61aaf52a85a0d8ed5dc`
- `git apply` the following patch:
```
diff --git a/CMakeLists.txt b/CMakeLists.txt
index ed3bc45..a8d83ab 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -402,11 +402,8 @@ ADD_CXX_FLAGS(-DYB_HEADERS_NO_STUBS=1)
 # For all builds:
 # For CMAKE_BUILD_TYPE=Debug
 #   -ggdb: Enable gdb debugging
-#   -O1: Enable minimum optimization
-#   -fno-omit-frame-pointer
-#       use frame pointers to allow simple stack frame walking for backtraces.
 # For CMAKE_BUILD_TYPE=FastDebug
-#   Same as DEBUG (TODO: remove fastdebug build type, use debug instead)
+#   Same as DEBUG, except with some optimizations on.
 # For CMAKE_BUILD_TYPE=Release
 #   -O3: Enable all compiler optimizations
 #   -g: Enable symbols for profiler tools (TODO: remove for shipping)
@@ -420,8 +417,8 @@ ADD_CXX_FLAGS(-DYB_HEADERS_NO_STUBS=1)
 #   2. Run the benchmarks (generates *.gcda profiling data).
 #   3. Build again with CMAKE_BUILD_TYPE_PROFILE_BUILD
 #     -fprofile-use: Compiler will use the profile outputs for optimizations
-set(CXX_FLAGS_DEBUG "-ggdb -O1 -fno-omit-frame-pointer")
-set(CXX_FLAGS_FASTDEBUG ${CXX_FLAGS_DEBUG})
+set(CXX_FLAGS_DEBUG "-ggdb")
+set(CXX_FLAGS_FASTDEBUG "-ggdb -O1 -fno-omit-frame-pointer")
 set(CXX_FLAGS_RELEASE "-O3 -g -DNDEBUG -fno-omit-frame-pointer")

 # Nullify CMake's predefined flags since we are handling different build types on our own.
```
- run `ybd --cxx-test pgwrapper_create_initial_sys_catalog_snapshot --gtest_filter CreateInitialSysCatalogSnapshotTest.CreateInitialSysCatalogSnapshot --clang11 -n 2 -- -p 1`

**Comment by ttyusupov** [comment]
```
[m-1] W0323 17:05:50.936890 21144 long_operation_tracker.cc:113] ScopedRWOperation running for 1.000s thread_id: 21535
[m-1] *** Aborted at 1616519150 (unix time) try "date -d @1616519150" if you are using GNU date ***
[m-1] PC: @                0x0 (unknown)
[m-1] *** SIGSEGV (@0x0) received by PID 21122 (TID 0x7f6c76edb700) from PID 0; stack trace: ***
[m-1]     @     0x7f6c87bdc5d0 (unknown)
[m-1]     @     0x7f6c89a5ac17  boost::atomics::detail::gcc_dcas_x86_64<false>::compare_exchange_strong(unsigned __int128 volatile&, unsigned __int128&, unsigned __int128, boost::memory_order, boost::memory_order) (/opt/yb-build/thirdparty/yugabyte-db-thirdparty-v20210306075844-471ef8e125-centos7-clang11/installed/uninstrumented/include/boost/atomic/detail/ops_gcc_x86_dcas.hpp:461)
[m-1]     @     0x7f6c89a5ac17  boost::atomics::detail::gcc_dcas_x86_64<false>::compare_exchange_weak(unsigned __int128 volatile&, unsigned __int128&, unsigned __int128, boost::memory_order, boost::memory_order) (/opt/yb-build/thirdparty/yugabyte-db-thirdparty-v20210306075844-471ef8e125-centos7-clang11/installed/uninstrumented/include/boost/atomic/detail/ops_gcc_x86_dcas.hpp:509)
[m-1]     @     0x7f6c89a5ac17  boost::atomics::detail::base_atomic<yb::LockFreeStack<yb::(anonymous namespace)::ThreadStackEntry>::Head, void>::compare_exchange_weak_impl(yb::LockFreeStack<yb::(anonymous namespace)::ThreadStackEntry>::Head&, yb::LockFreeStack<yb::(anonymous namespace)::ThreadStackEntry>::Head const&, boost::memory_order, boost::memory_order, std::integral_constant<bool, true>) volatile (/opt/yb-build/thirdparty/yugabyte-db-thirdparty-v20210306075844-471ef8e125-centos7-clang11/installed/uninstrumented/include/boost/atomic/detail/atomic_template.hpp:281)
[m-1]     @     0x7f6c89a5ac17  boost::atomics::detail::base_atomic<yb::LockFreeStack<yb::(anonymous namespace)::ThreadStackEntry>::Head, void>::compare_exchange_weak(yb::LockFreeStack<yb::(anonymous namespace)::ThreadStackEntry>::Head&, yb::LockFreeStack<yb::(anonymous namespace)::ThreadStackEntry>::Head const&, boost::memory_order, boost::memory_order) volatile (/opt/yb-build/thirdparty/yugabyte-db-thirdparty-v20210306075844-471ef8e125-centos7-clang11/installed/uninstrumented/include/boost/atomic/detail/atomic_template.hpp:249)
[m-1]     @     0x7f6c89a5ac17  boost::atomics::detail::base_atomic<yb::LockFreeStack<yb::(anonymous namespace)::ThreadStackEntry>::Head, void>::compare_exchange_weak(yb::LockFreeStack<yb::(anonymous namespace)::ThreadStackEntry>::Head&, yb::LockFreeStack<yb::(anonymous namespace)::ThreadStackEntry>::Head const&, boost::memory_order) volatile (/opt/yb-build/thirdparty/yugabyte-db-thirdparty-v20210306075844-471ef8e125-centos7-clang11/installed/uninstrumented/include/boost/atomic/detail/atomic_template.hpp:254)
[m-1]     @     0x7f6c89a5ac17  yb::LockFreeStack<yb::(anonymous namespace)::ThreadStackEntry>::Pop() (src/yb/util/lockfree.h:128)
[m-1]     @     0x7f6c89a5a559  yb::(anonymous namespace)::ThreadStackHelper::RecordStackTrace(yb::StackTrace const&) (src/yb/util/debug-util.cc:255)
[m-1]     @     0x7f6c89a5a507  yb::(anonymous namespace)::HandleStackTraceSignal(int) (src/yb/util/debug-util.cc:279)
[m-1]     @     0x7f6c87836280 (unknown)
```

**Comment by sanketkedia** [comment]
Tried reproducing this with boost::atomic as well as with std::atomic. It had reproduced with boost::atomic but worked fine with std::atomic. This might be a boost<>clang version mismatch. Boost version was 1.69 and clang is 11.0.0. The platform is centos (Linux dev-server-sanket 3.10.0-1127.19.1.el7.x86_64 #1 SMP Tue Aug 25 17:23:54 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux). Maybe, we could try forcing a particular version of boost > 1.69.

**Comment by bmatican** [comment]
@spolitov @ttyusupov any reason we're using boost here instead of std? if not, then might as well switch away from boost

**Comment by ttyusupov** [comment]
@spolitov , can we switch `LockFreeStack` to std::atomic or any reasons not to do it?

**Comment by spolitov** [comment]
Because std::atomic is not lock-free.
Also I suppose this issue was fixed by D12071/ff5ee81a128b1b1bb4f00e84275cb6c4052a2755
